### PR TITLE
feat(backend/frontend/shared): add search functionality to friends page tm-561

### PR DESCRIPTION
**Pull request by Sytnikovakatya**
Made "Search" field on"Friends" page to perform a search among the users present on the page

https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-trackmates/assets/106179834/d86598ca-320e-4ecf-8de2-0013e6c98968

### CODE REVIEW

File: apps/backend/src/modules/friends/friend.controller.ts [hunk 1]
```diff
@@ -526,13 +530,13 @@ class FriendController extends BaseController {
 	 *                 $ref: "#/components/schemas/User"
 	 */
 	private async getPotentialFollowers({
-		query: { count, page },
+		query: { count, page, search },
 		user: { id },
 	}: APIHandlerOptions<{
-		query: PaginationRequestDto;
+		query: { search: string | undefined } & PaginationRequestDto;
```
Comment (Ann-Maksay) [inline apps/backend/src/modules/friends/friend.controller.ts:536]: Why might it be undefined? We already do this check and send `search: searchQuery ? ""` from the frontend

Comment (Sytnikovakatya) [reply apps/backend/src/modules/friends/friend.controller.ts:536]: I added to check when we use, for example, Swagger. And I also saw  the same checking  in the user course controller

File: apps/frontend/src/modules/friends/friends-api.ts [hunk 2]
```diff
@@ -55,7 +55,7 @@ class FriendsApi extends BaseHTTPApi {
 	}
 
 	public async getAllPotentialFriends(
-		query: PaginationRequestDto,
+		query: { search: string } & PaginationRequestDto,
```
Comment (what1s1ove) [inline apps/frontend/src/modules/friends/friends-api.ts:58]: you duplicate this piece `{ search: string } & PaginationRequestDto` a lot. create a separate type for it

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/friends/friend.controller.ts [hunk 3]
```diff
@@ -137,7 +137,7 @@ class FriendController extends BaseController {
 			handler: (options) => {
 				return this.getUserFollowers(
 					options as APIHandlerOptions<{
-						query: PaginationRequestDto;
+						query: { search: string } & PaginationRequestDto;
```
Comment (what1s1ove) [inline apps/backend/src/modules/friends/friend.controller.ts:140]: use a new type in all places where you have added `{ search: string }`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

**Comment by Vfivew** [comment]
Test it with pagination hook.

File: packages/shared/src/modules/friends/libs/types/friend-list-request-dto.type.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { type PaginationRequestDto } from "../../../../libs/types/types.js";
+
+type FriendListRequestDto = {
```
Comment (fshabanov) [inline packages/shared/src/modules/friends/libs/types/friend-list-request-dto.type.ts:3]:
```suggestion
type FriendListRequestQueryDto = {
```

**Review by fshabanov — CHANGES_REQUESTED** [review]

File: apps/backend/src/modules/friends/friend.controller.ts [hunk 5]
```diff
@@ -550,6 +557,10 @@ class FriendController extends BaseController {
 	 *     security:
 	 *       - bearerAuth: []
 	 *     parameters:
+	 * 		- name: search
+	 * 		   in: query
```
Comment (fshabanov) [inline apps/backend/src/modules/friends/friend.controller.ts:561]:
```suggestion
	 * 		- name: search
	 * 		in: query
```
In other places we have such indentation

**Review by fshabanov — CHANGES_REQUESTED** [review]

**Review by Oleksandra-Masiuk — APPROVED** [review]

**Review by fshabanov — APPROVED** [review]

**Review by what1s1ove — APPROVED** [review]

**Pull request merged**